streamer/pulp/streamer/cache.py

```diff
@@ -0,0 +1,167 @@
+import sys
+
+from logging import getLogger
+from threading import RLock
+from datetime import datetime, timedelta
+
+log = getLogger(__name__)
+
+
+class NotCached(Exception):
+    """
+    Requested object is not found in the cache.
+    """
+    pass
+
+
+class Cache(object):
+    """
+    Generic object cache.
+
+    Attributes:
+        eviction_threshold (timedelta): How long an unrequested item will be cached.
+        _lock (RLock): The object mutex.
+        _inventory (dict): The inventory of cached objects.
+            Each value is an Item.
+    """
+
+    def __init__(self, eviction_threshold=None):
+        """
+        Args:
+            eviction_threshold (timedelta): How long an unrequested item will be cached.
+
+        """
+        self.eviction_threshold = eviction_threshold or timedelta(hours=4)
+        self._lock = RLock()
+        self._inventory = {}
+
+    def add(self, key, object):
+        """
+        Add an object to the cache.
+
+        Args:
+            key (hashable): The caching key.
+            object (object): An object to be cached.
+        """
+        with self._lock:
+            self._inventory[key] = Item(object)
+
+    def purge(self, key):
+        """
+        Purge (delete) objects cached using the specified key.
+
+        Args:
+            key (hashable): The caching key.
+        """
+        with self._lock:
+            return self._inventory.pop(key)
+
+    def get(self, key):
+        """
+        Get a cached object by key.
+
+        Args:
+            key (hashable): The caching key.
+
+        Returns:
+            object: The requested cached object.
+
+        Raises:
+            NotCached: When not found in the cache.
+        """
+        with self._lock:
+            try:
+                item = self._inventory[key]
+            except KeyError:
+                raise NotCached()
+            item.touch()
+            self.evict(self.eviction_threshold)
+            return item.object
+
+    def evict(self, threshold):
+        """
+        Evict all unused cached objects.
+
+        Args:
+            threshold (timedelta): Evict all cached objects that have not been
+                requested since the specified timedelta.
+
+        Returns:
+            list: The evicted objects.
```

Comment:
The intention here was to support the orderly destruction of evicted objects.  For example, if the evicted session needed to be closed, the subclass in the streamer could override (call super as it should) and call Session.close() on each evicted session.  The missing test is an oversight.